src/play_controller.cpp

```diff
@@ -280,7 +281,59 @@ void play_controller::init(const config& level)
 		plugins_context_->set_callback("save_game", [this](const config& cfg) { save_game_auto(cfg["filename"]); }, true);
 		plugins_context_->set_callback("save_replay", [this](const config& cfg) { save_replay_auto(cfg["filename"]); }, true);
 		plugins_context_->set_callback("quit", [](const config&) { throw_quit_game_exception(); }, false);
+		plugins_context_->set_callback_execute(*resources::lua_kernel, true);
 		plugins_context_->set_accessor_string("scenario_name", [this](config) { return get_scenario_name(); });
+		plugins_context_->set_accessor_int("current_side", [this](config) { return current_side(); });
+		plugins_context_->set_accessor_int("current_turn", [this](config) { return turn(); });
+		plugins_context_->set_accessor_bool("can_move", [this](config) { return !events::commands_disabled && gamestate().gamedata_.phase() == game_data::TURN_PLAYING; });
+		plugins_context_->set_callback("end_turn", [this](config) { force_end_turn(); }, false);
+		plugins_context_->set_callback("synced_command", [this](config cmd) {
+			auto& pm = *plugins_manager::get();
+			if(resources::whiteboard->has_planned_unit_map())
+			{
+				ERR_NG << "plugin called synced command while whiteboard is applied, ignoring";
+				pm.notify_event("synced_command_error", config{"error", "whiteboard"});
+				return;
+			}
+
+			auto& gamedata = gamestate().gamedata_;
+			const bool is_too_early = gamedata.phase() == game_data::INITIAL || resources::gamedata->phase() == game_data::PRELOAD;
+			const bool is_during_turn = gamedata.phase() == game_data::TURN_PLAYING;
+			const bool is_unsynced = synced_context::get_synced_state() == synced_context::UNSYNCED;
+			if(is_too_early) {
+				ERR_NG << "synced command called too early, only allowed at START or later";
+				pm.notify_event("synced_command_error", config{"error", "too-early"});
+				return;
+			}
+			if(is_unsynced && !is_during_turn) {
+				ERR_NG << "synced command can only be used during a turn when a user would also be able to invoke commands";
+				pm.notify_event("synced_command_error", config{"error", "not-your-turn"});
+				return;
+			}
+			if(is_unsynced && events::commands_disabled) {
+				ERR_NG << "synced command cannot be invoked while commands are blocked";
+				pm.notify_event("synced_command_error", config{"error", "disabled"});
+				return;
+			}
+			if(is_unsynced && !resources::controller->current_team().is_local()) {
+				ERR_NG << "synced command can only be used from clients that control the currently playing side";
+				pm.notify_event("synced_command_error", config{"error", "not-your-turn"});
+				return;
+			}
```

Comment:
this whole block was coped from the [do_command] implementation i assume, it woudl be good if we could factor out this code so that is only exists once

Reply:
Yeah, it was _mostly_ copied, but with two significant differences:

* The log messages don't refer to `[do_command]`
* It's not restricted in which commands it can execute